Expose NavigationMeshGenerator API for custom modules / addons / extentions

smix8 opened the pull request:
Implementation draft for proposal https://github.com/godotengine/godot-proposals/issues/5138 to open the `NavigationMeshGenerator` API and `NavigationMesh` baking for custom modules / addons / extentions and move the node type specific code to dedicated files.

Adds `NavigationGeometryParser3D` class.

This class can be extended and registered on the NavigationMeshGenerator singleton to contribute 3D geometry from specified node class(es) for navmesh baking.
```gdscript
var parser : NavigationGeometryParser3D = NavigationGeometryParser3D.new()
NavigationMeshGenerator.register_geometry_parser_3d(parser)
NavigationMeshGenerator.unregister_geometry_parser_3d(parser)
```
Modules can use the `parses_node(`) and `parse_geometry()` directly while scripts can override the underscore `_parses_node()` and `_parse_geometry()` functions. Since navmesh baking with ReCast requires specific, triangulated mesh data helper functions are exposed to add mesh resources or mesh arrays or single faces to the source geometry.

GDScript example with function override:

```gdscript
extends NavigationGeometryParser3D

func _parses_node(node : Node):
    # return true if this node wants to add baking geometry
    return node is CustomNode3D

func _parse_geometry(node: Node, navigationmesh : NavigationMesh):
    var mesh : Mesh = PlaneMesh.new()
    mesh.size = Vector2(10.0,10.0)
    var mesh_xfrom : Transform3D = node.global_transform
    add_mesh(mesh, mesh_xfrom)
    
    var mesh_array : Array = mesh.get_mesh_arrays()
    mesh_xfrom.origin.x += 5.0
    mesh_xfrom.origin.z += 5.0
    add_mesh_array(mesh_array, mesh_xfrom)
    
    mesh_xfrom.origin.x -= 10.0
    mesh_xfrom.origin.z -= 10.0
    var faces : PackedVector3Array = mesh.get_faces()
    add_faces(faces, mesh_xfrom)
```

<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.
-->

smix8 committed:
Expose NavigationMeshGenerator API for custom addons

Adds NavigationGeometryParser3D class that contributes 3D geometry from specified node class(es) for navmesh baking.

Zylann reviewed (commented):
That looks good! Pretty much what I thought.

I was thinking the arguments could have been an object, and helper functions would also have been on that object, which allows to add more things in the future without breaking ABI, and also make the geometry generator stateless regarding its function (cuz it feels odd for it to call functions on itself to affect data that is actually "passed to it" rather. Could even be threaded). But it's not an absolute requirement, more like a design idea.

smix8 commented on modules/navigation/navigation_geometry_parser_3d.h:
```diff
@@ -0,0 +1,69 @@
+
+public:
+	virtual bool parses_node(Node *p_node);
+	virtual void parse_geometry(const Transform3D &p_navmesh_transform, Node *p_node, NavigationMesh::ParsedGeometryType p_generate_from, uint32_t p_collision_mask);
```
`navmesh_transform` is the source transform from the start of the baking process. NavigationMesh resources have no transform so every offset needs to be baked. Maybe another term could make the parameter more clear?

Zylann replied:
What I mean is, you used it in every single call to functions adding geometry, which suggests it's something the navmesh generator can do by itself, since not applying it would be wrong anyways. Is there any use case that would use it in any different way?
(I'm not referring to the transform of individual nodes providing geometries, these should still be provided)

smix8 replied:
Properly the generator can handle that part, that was just how it was before. I only removed parameters from functions that didn't work with scripting, like passing the vertices and indices array's with every function call.

Zylann replied:
In pseudocode:
```cpp
// Node X says "here's my geometry, and here is where I am"
add_arrays(arrays, transform)

// The geometry parser knows which navmesh it is processing,
// so it can combine its transform internally
void add_arrays(Arrays geometry, Transform3D geom_transform) {
	Transform3D total_transform = navmesh_transform * geom_transform;
	// Apply to geometry and add to navmesh
}
```

smix8 commented on modules/navigation/geometry_parser/gridmap_navigation_geometry_parser_3d.h:
```diff
@@ -0,0 +1,192 @@
+							PackedVector3Array vertices = data;
+							Geometry3D::MeshData md;
+
+							Error err = ConvexHullComputer::convex_hull(vertices, md);
```
I don't know, not familiar with that part in Godot, all this parse code was a 1:1 copy from the original parse source.

smix8 commented on modules/navigation/geometry_parser/staticbody3d_navigation_geometry_parser_3d.h:
```diff
@@ -0,0 +1,183 @@
+						}
+
+						HeightMapShape3D *heightmap_shape = Object::cast_to<HeightMapShape3D>(*s);
+						if (heightmap_shape) {
```
It is duplicated cause the gridmap module can be disabled and adding a helper function just for the gridmap felt off from how the rest of the parsing was handled so I didn't touch it and followed through.

smix8 commented on modules/navigation/navigation_geometry_parser_3d.h:
```diff
@@ -0,0 +1,69 @@
+	GDVIRTUAL4C(_parse_geometry, const Transform3D &, Node *, NavigationMesh::ParsedGeometryType, uint32_t);
+
+private:
+	Vector<float> *total_parsed_vertices = nullptr;
```
As far as I remember this is was ReCast expects when the parsed geometry is handed over for the baking.

smix8 commented on modules/navigation/doc_classes/NavigationGeometryParser3D.xml:
```diff
@@ -0,0 +1,54 @@
+			<param index="1" name="xform" type="Transform3D" />
+			<description>
+				Adds an array of vertex positions to the geometry data for navmesh baking to form triangulated faces. For each face the array must have three vertex positions in clockwise winding order. Since [NavigationMesh] resource have no transform all vertex positions need to be offset by the node's transform using the [code]xform[/code] parameter.
+			</description>
```
Yes the format that ReCast expects is an indexed mesh with CCW winding order, all the helper functions do the conversion in the background from the Godot Mesh format. This function was added because some stuff in Godot (physics) returns no proper mesh format but just an array of positions and this function calls _add_faces() which converts it to proper vertex / index array format.

Zylann commented:
From a discussion on RocketChat, it appears the NavigationServer singleton isn't available until all initialization levels of all the modules have finished to run. If that's true, it means this feature can't possibly work. Can you confirm?
That's quite bad though, it looks like an oversight. Initialization levels were created to deal with this kind of problem, but if NavigationServer doesn't use it, that defeats the whole point here^^"

smix8 commented:
Changed how the navmesh transform is handled as suggested and removed the individual parameters in favor of the NavigationMesh ref so all baking properties are available not just the source geometry typ and collision layer.

Not sure what the GCC sanitizer tests build check wants from me or why it is complaining when the array with the parser RefCount objects is cleared. Cannot reproduce any of this local.

DarkKilauea commented on modules/navigation/navigation_geometry_parser_3d.h:
```diff
@@ -0,0 +1,70 @@
+
+	GDVIRTUAL1RC(bool, _parses_node, Node *);
+
+	GDVIRTUAL2C(_parse_geometry, Node *, Ref<NavigationMesh>);
```
Right now, parsers need to check the `parsed_geometry_type ` to know if they should be looking for collision shapes or visual meshes.

What if instead, we have two methods here:
- `_parse_geometry` (would parse visual geometry)
- `_parse_collision` (would parse only collision shapes)

Then, if a parser supports only visual meshes, it would implement `_parse_geometry`.  If it only supports collision shapes, it would implement `_parse_collision`.  If it supports both, it implements both.

This should simplify the implementation of parsers and might remove the need to pass in the target `NavigationMesh` at all.  The base parser would be in charge of calling the correct method based on the `parsed_geometry_type` field.

smix8 replied:
I don't think we can stop passing the NavigationMesh ref as long as this object holds the user set bake properties. Users might want to change the added source geometry entirely depending on some of those values e.g. snap to cell value or simplify depending on slope value.

While most of the code that checks for parsed_geometry_type was a copy&paste from existing parsing I am not sure adding this function distinction between collision and everything else has any real benefit. The source geometry parser does not care what kind of object adds source geometry, could be just a basic node that does it, so this distinction would be entirely artifical. Having a single function for everything keeps things simple and allows addon / module devs to do whatever they want. When you have specific functions and someone has a shared node class that wants to add both visuals and collisions in a very specific order you are already creating obstacles for them by splitting functions with a fixed order for nothing.

DarkKilauea commented on modules/navigation/geometry_parser/gridmap_navigation_geometry_parser_3d.h:
```diff
@@ -0,0 +1,195 @@
+#include "scene/resources/sphere_shape_3d.h"
+#include "scene/resources/world_boundary_shape_3d.h"
+
+class GridMap3DNavigationGeometryParser3D : public NavigationGeometryParser3D {
```
This should probably be implemented in the `gridmap` module so we don't have to check if the module is active.

smix8 replied:
Agree but I think first the NavigationMeshGenerator needs to move from the navigation module or at least the geometry parser needs to move e.g. to server/navigation or scene/main so we don't have an odd dependency in the other direction.

DarkKilauea commented on modules/navigation/geometry_parser/csgshape3d_navigation_geometry_parser_3d.h:
```diff
@@ -0,0 +1,59 @@
+#include "modules/csg/csg_shape.h"
+#include "modules/navigation/navigation_geometry_parser_3d.h"
+
+class CSGShape3DNavigationGeometryParser3D : public NavigationGeometryParser3D {
```
This should probably be implemented in the `csg` module so we don't have to check if the module is active.

smix8 replied:
same as above

DarkKilauea commented on modules/navigation/register_types.cpp:
```diff
@@ -59,24 +70,38 @@ void initialize_navigation_module(ModuleInitializationLevel p_level) {
+		GDREGISTER_CLASS(NavigationGeometryParser3D);
+		// add default 3D node navigation geometry parsers
+#ifdef MODULE_GRIDMAP_ENABLED
+		NavigationMeshGenerator::get_singleton()->register_geometry_parser_3d(memnew(GridMap3DNavigationGeometryParser3D));
```
The registrations of the parsers should also ideally happen in the modules that they are needed for, to avoid every module needing an `ifdef` here.

smix8 replied:
agree

smix8 commented on modules/navigation/navigation_mesh_generator.cpp:
```diff
@@ -161,311 +161,18 @@ void NavigationMeshGenerator::_add_faces(const PackedVector3Array &p_faces, cons
-		}
-	}
-#endif
+void NavigationMeshGenerator::_parse_geometry_node(const Transform3D &p_navmesh_transform, Ref<NavigationMesh> p_navigationmesh, Node *p_node, Vector<float> &p_vertices, Vector<int> &p_indices, bool p_recurse_children) {
```
I don't really want to make the generic base parser dependent on any specific SceneTree object. The only reason "add_mesh()" exists is cause it works with a generic, triangulated and indexed mesh array not with a specific mesh primitive. On top the physics shapes have often very weird optimized polygon formats that need very specific conversions to get them into a valid mesh format, better imo to keep that hazard contained inside specific parser code.

smix8 commented on modules/navigation/navigation_mesh_generator.cpp:
```diff
@@ -161,311 +161,18 @@ void NavigationMeshGenerator::_add_faces(const PackedVector3Array &p_faces, cons
-#endif
+void NavigationMeshGenerator::_parse_geometry_node(const Transform3D &p_navmesh_transform, Ref<NavigationMesh> p_navigationmesh, Node *p_node, Vector<float> &p_vertices, Vector<int> &p_indices, bool p_recurse_children) {
+	// find a geometry parser that handles the node
+	for (uint32_t i(0); i < NavigationMeshGenerator::get_singleton()->geometry_3d_parsers.size(); i++) {
```
I don't know what happened.

smix8 commented:
Superseded by https://github.com/godotengine/godot/pull/70724
